packages/flutter/lib/src/material/icon.dart

```diff
@@ -65,13 +72,39 @@ class Icon extends StatelessComponent {
       category = parts[0];
       subtype = parts[1];
     }
-    // TODO(eseidel): This clearly isn't correct.  Not sure what would be.
-    // Should we use the ios images on ios?
-    String density = 'drawable-xxhdpi';
-    String colorSuffix = _getColorSuffix(context);
+    String density = new DensityQualifier(context).resolve();
+    String colorSuffix = new ColorQualifier(context: context, color: color).resolve();
+    return new Future.sync(() => '$category/drawable-$density/ic_${subtype}_${colorSuffix}_${iconSize}dp.png');
+  }
+}
+
+class Icon extends StatelessComponent {
+  Icon({
+    Key key,
+    this.size: IconSize.s24,
+    this.icon: '',
+    this.color,
+    this.colorFilter
+  }) : super(key: key) {
+    assert(size != null);
+    assert(icon != null);
+  }
+
+  final IconSize size;
+  final String icon;
+  final IconThemeColor color;
+  final ColorFilter colorFilter;
+
+  Widget build(BuildContext context) {
+    MaterialIconResolver resolver = new MaterialIconResolver(
+      context: context,
+      iconSize: _kIconSize[size],
+      color: color
+    );
     int iconSize = _kIconSize[size];
     return new AssetImage(
-      name: '$category/$density/ic_${subtype}_${colorSuffix}_${iconSize}dp.png',
+      name: icon,
+      resolver: resolver,
```

Comment:
I'd inline the resolver creation to make it more obvious when reading the AssetImage resolver that the resolver is not cached or something.